legacy/common.c

```diff
@@ -83,7 +84,17 @@ void __assert_func(const char *file, int line, const char *func,
 }
 #endif
 
-void hal_delay(uint32_t ms) { usbSleep(ms); }
+void hal_delay(uint32_t ms) {
+#if EMULATOR
+  usleep(ms * 1000);
```

Comment:
The reason why we used usbSleep here is that device responds to usb packets when it is sleeping.
Are we 100% sure this will not cause any problems (device randomly disconnecting)?

Reply:
With `usbSleep` device replies to *USB protobuf* messages.

Even with NOP loop (like in the PR), the USB itself works and doesn't cause disconnects, USB IRQ handlers put the data in FIFO. That `hal_delay` is only called when PIN is checked and user is "punished" for bad attempts (where no protobuf messages should be handled). Once normal operation is resumed, USB protobuf data is handled normally.

I've tried it on hardware T1 with Suite doing setup several times, also tried to change PIN, did a lot of wrong PIN inputs to increase the delay up to 16 seconds.

Everything works as it did before as far as I can tell.

Reply:
Thanks!

Reply:
So far I've discovered that having `hal_delay` as NOP should work with legitimate clients (no recursion attempts which should not work anyway), malicious clients could just send bad PIN attempts to try device lockout anyway even now. However it's hard to try out all race conditions with client/T1 disconnects over all OSes to be 100% sure there is no edge case not covered (experimentally everything works as it's supposed even when trying disconnects, but...). So I'd probably have `hal_delay` to be a NOP sleep, but explicitly use the `waitAndProcessUSBRequests` in unlock as to not appear to be a no-op delay.

Reply:
After careful code analysis, `hal_delay` with NOP should work fine and I did test various race conditions induced manually as well. If busy loops broke USB low-level responsiveness, it would have also happened in long signing operations, for example, as USB low-level timeouts are very short.